nknd.go

```diff
@@ -375,3 +384,112 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+func GetChordID(seeds []string, publickey []byte) ([]byte, error) {
+	id, err := chain.DefaultLedger.Store.GetID(publickey)
+	if err == nil {
+		if len(id) != 0 && !bytes.Equal(id, make([]byte, 32)) {
+			return id, nil
+		}
+	}
+
+	log.Warningf("get chord ID from localnode faild, no chord ID in ledger")
+
+	rand.Shuffle(len(seeds), func(i int, j int) {
+		seeds[i], seeds[j] = seeds[j], seeds[i]
+	})
+
+	for _, seed := range seeds {
+		id, err := client.GetChordID(seed, publickey)
+		if err == nil {
+			return id, nil
+		}
+
+		log.Warningf("get chordID from %s met error: %v", seed, err)
+
+	}
+
+	return nil, errors.New("get chordID failed")
+}
+
+func CreateChordID(seeds []string, wallet vault.Wallet, regFee Fixed64) error {
+	account, err := wallet.GetDefaultAccount()
+	if err != nil {
+		return err
+	}
+
+	addr, err := account.ProgramHash.ToAddress()
+	if err != nil {
+		return err
+	}
+
+	rand.Shuffle(len(seeds), func(i int, j int) {
+		seeds[i], seeds[j] = seeds[j], seeds[i]
+	})
+
+	for _, seed := range seeds {
+		nonce, err := client.GetNonceByAddr(seed, addr)
+		if err != nil {
+			log.Warningf("get nonce from %s met error: %v", seed, err)
+			continue
+		}
+
+		txn, err := common.MakeGenerateIDTransaction(wallet, regFee, nonce, 0)
+		if err != nil {
+			return err
+		}
+
+		buff, err := txn.Marshal()
+		if err != nil {
+			return err
+		}
+
+		_, err = client.CreateChordID(seed, hex.EncodeToString(buff))
+		if err != nil {
+			log.Warningf("create chordID from %s met error: %v", seed, err)
+			continue
+		}
+
+		return nil
+	}
+
+	return errors.New("create ChordID failed")
+}
+
+func GetOrCreateChordID(seeds []string, wallet vault.Wallet, regFee Fixed64) ([]byte, error) {
+	account, err := wallet.GetDefaultAccount()
+	if err != nil {
+		return nil, err
+	}
+	pkEncoded, err := account.PubKey().EncodePoint(true)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := GetChordID(seeds, pkEncoded)
+	if err == nil {
+		return id, nil
+	}
+
+	log.Warningf("chord ID is not found in local db or remote node: %v", err)
+
+	if err := CreateChordID(seeds, wallet, regFee); err != nil {
+		return nil, err
+	}
+
+	timeout := time.After(time.Minute * 3)
+out:
+	for {
+		select {
+		case <-timeout:
+			break out
+		case <-time.After(time.Second * 40):
```

Comment:
Instead of hard code time here, we should wait 2 * config.ConsensusDuration initially, and then check every config.ConsensusDuration